clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -56,10 +56,12 @@ public int sizeInBytes() {
 
     @Override
     public long writeTo(GatheringByteChannel channel, long position, int length) throws IOException {
-        ByteBuffer dup = buffer.duplicate();
+        if (position > Integer.MAX_VALUE)
+            throw new IllegalArgumentException("position should not be larger than Integer.MAX_VALUE: " + position);
         int pos = (int) position;
+        ByteBuffer dup = buffer.duplicate();
         dup.position(pos);
-        dup.limit(pos + length);
+        dup.limit(Math.min(pos + length, buffer.limit()));
```

Comment:
Is this right? If the `pos + length` is bigger than the `buffer.limit()` doesn't that indicate an invalid write? In other words, instead of setting the limit on the duplicated buffer, shouldn't there be a check before this that `pos + length <= buffer.limit()`?